plugins/languages/treesitter/treesitter.nix

```
@@ -6,210 +6,253 @@
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @lua)
-              (indented_string_expression (string_fragment) @lua)
-            ]
-            (#match? @_path "^extraConfigLua(Pre|Post)?$"))
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @vim)
-              (indented_string_expression (string_fragment) @vim)
-            ]
-            (#match? @_path "^extraConfigVim(Pre|Post)?$"))
+          This is a wrapping of the `vim.treesitter.language.register` function.
+          Register specific parsers to one or several filetypes.
+          The keys are the parser names and the values are either one or several filetypes.
         '';
-      };
 
-      extraPlugins =
-        if cfg.nixGrammars then [ (cfg.package.withPlugins (_: cfg.grammarPackages)) ] else [ cfg.package ];
-      extraPackages = with pkgs; [
-        tree-sitter
-        nodejs
-        cfg.gccPackage
-      ];
+    nixGrammars = mkOption {
+      type = types.bool;
+      default = true;
+      description = "Install grammars with Nix";
+    };
 
-      opts = mkIf cfg.folding {
-        foldmethod = "expr";
-        foldexpr = "nvim_treesitter#foldexpr()";
-      };
+    nixvimInjections = helpers.defaultNullOpts.mkBool true ''nixvim specific injections, like lua highlighting in extraConfigLua'';
+  };
+
+  extraConfig = cfg: {
+    extraConfigLua =
+      (optionalString (cfg.settings.parser_install_dir != null) ''
+        vim.opt.runtimepath:append("${cfg.settings.parser_install_dir}")
+      '')
+      + ''
+        require('nvim-treesitter.configs').setup(${helpers.toLuaObject cfg.settings})
+      ''
+      + (optionalString (cfg.languageRegister != { }) ''
+        __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}
+
+        for parser_name, ft in pairs(__parserFiletypeMappings) do
+          require('vim.treesitter.language').register(parser_name, ft)
+        end
+      '');
+
+    extraFiles = mkIf cfg.nixvimInjections {
+      "queries/nix/injections.scm" = ''
+        ;; extends
+
+        (binding
+          attrpath: (attrpath (identifier) @_path)
+          expression: [
+            (string_expression (string_fragment) @lua)
+            (indented_string_expression (string_fragment) @lua)
+          ]
+          (#match? @_path "^extraConfigLua(Pre|Post)?$"))
+
+        (binding
+          attrpath: (attrpath (identifier) @_path)
+          expression: [
+            (string_expression (string_fragment) @vim)
+            (indented_string_expression (string_fragment) @vim)
+          ]
+          (#match? @_path "^extraConfigVim(Pre|Post)?$"))
+      '';
     };
+
+    extraPlugins = mkMerge [
+      (mkIf cfg.nixGrammars [ (cfg.package.withPlugins (_: cfg.grammarPackages)) ])
+      (mkIf (!cfg.nixGrammars) [ cfg.package ])
+    ];
```

We will need to add an option to `mkNeovimPlugin` to disable it also installing `cfg.package` (similar to options like `callSetup`). I can do that in a separate PR.

---

Also wondering why it was done this way in the derivation... it seems like the intent is just to override the package with the withPlugins when we have nixGrammars enabled. The package being included when false seems confusing since that'll already be provided by the cfg.package with the package option. Would it make more sense to just do this?

```suggestion
    extraPlugins = optionals cfg.nixGrammars [ (cfg.package.withPlugins (_: cfg.grammarPackages)) ];
```

---

> The package being included when false seems confusing since that'll already be provided by the cfg.package with the package option

The intention is to install _either_ `cfg.package` as-is, _or_ install a modified version of `cfg.package` that bundles the `grammarPackages`.

Your suggestion would end up installing _two_ treesitter packages when `cfg.nixGrammars` is true. This is also why we need to disable `mkNeovimPlugin` from installing the package.

---

As a side note:

It's _usually_ better to use `mkIf` instead of `optionals` in module configs

This is because `mkIf` returns a `{ _type = "if"; condition = true/false; content = []; }` wrapper that the module system can merge into the final config if needed, but it can still validate `content` is correct even when `condition = false`. This approach can also avoid some inf-recursive conditions that you'd run into with a "real" nix `if then else` expression.

But sometimes `if then else` is more readable:

```nix
extraPlugins = [
  (if cfg.nixGrammars then cfg.package.withPlugins (_: cfg.grammarPackages) else cfg.package)
]
```

---

Should also be wrapped in a `mkIf (cfg.package != null)`.

We probably also want assertions for any options that need a non-null cfg.package being used.